core/libp2p/transport/tcp/tcp_listener.hpp

```diff
@@ -17,124 +17,30 @@ namespace libp2p::transport {
-      using namespace boost::asio;  // NOLINT
-      try {
-        OUTCOME_TRY(endpoint, detail::makeEndpoint(address));
+    outcome::result<multi::Multiaddress> getListenMultiaddr() const override;
 
-        // setup acceptor, throws
-        acceptor_.open(endpoint.protocol());
-        acceptor_.set_option(ip::tcp::acceptor::reuse_address(true));
-        acceptor_.bind(endpoint);
-        acceptor_.listen();
+    bool isClosed() const override;
 
-        // start listening
-        do_accept();
-
-        return outcome::success();
-      } catch (const boost::system::system_error &e) {
-        return e.code();
-      }
-    };
-
-    bool canListen(const multi::Multiaddress &ma) const override {
-      return detail::supportsIpTcp(ma);
-    };
-
-    outcome::result<multi::Multiaddress> getListenMultiaddr() const override {
-      return detail::makeEndpoint(acceptor_.local_endpoint());
-    };
-
-    bool isClosed() const override {
-      return !acceptor_.is_open();
-    };
-
-    outcome::result<void> close() override {
-      boost::system::error_code ec;
-      acceptor_.close(ec);
-      return ec;
-    };
+    outcome::result<void> close() override;
 
    private:
+    boost::asio::io_context& context_;
     boost::asio::ip::tcp::acceptor acceptor_;
     std::shared_ptr<Upgrader> upgrader_;
     TransportListener::HandlerFunc v_;
-    TransportListener::ErrorFunc e_;
-
-    void do_accept() {
-      using namespace boost::asio;    // NOLINT
-      using namespace boost::system;  // NOLINT
-
-      if (!acceptor_.is_open()) {
-        return;
-      }
-
-      acceptor_.async_accept([self{this->shared_from_this()}](
-                                 const boost::system::error_code &ec,
-                                 ip::tcp::socket sock) {
-        if (ec) {
-          return self->e_(ec);
-        }
-
-        ufiber::spawn(
-            self->acceptor_.get_executor(),
-            [sock{std::move(sock)},
-             self{std::move(self)}](auto &&yield) mutable {
-              using E = decltype(yield.get_executor());
-              auto conn =
-                  std::make_shared<TcpConnection<E>>(yield, std::move(sock));
-
-              // upgrade to secure
-              auto rsc = self->upgrader_->upgradeToSecure(std::move(conn));
-              if (!rsc) {
-                return self->e_(rsc.error());
-              }
-
-              // upgrade to muxed
-              auto rmc =
-                  self->upgrader_->upgradeToMuxed(std::move(rsc.value()));
-              if (!rmc) {
-                return self->e_(rmc.error());
-              }
-
-              // execute user-provided callback
-              auto r = self->v_(std::move(rmc.value()));
-              if (!r) {
-                // signal error in case of failure
-                return self->e_(r.error());
-              }
-            });
 
-        self->do_accept();
-      });
-    }
+    void do_accept();
```

Comment:
Please, docs. Otherwise, it can be not clear that this starts an accept loop

Reply:
it is private method, it should not bother you :)